# [5주차 기본과제] 뉴스 기사 분류를 Gemma로 구현하기


In [88]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")
tokenizer = AutoTokenizer.from_pretrained("google/gemma-2-2b-it")
model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2-2b-it",
    device_map="auto",
    torch_dtype=torch.bfloat16,
)

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: fineGrained).
Your token has been saved to /Users/obov/.cache/huggingface/token
Login successful


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/47.0k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.24M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/636 [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [89]:
from datasets import load_dataset


news_ds = load_dataset("fancyzhx/ag_news")
news_ds

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 7600
    })
})

In [90]:
world_ds = news_ds["train"].filter(lambda data: data["label"] == 0)
world_ds["text"][:3]

['Venezuelans Vote Early in Referendum on Chavez Rule (Reuters) Reuters - Venezuelans turned out early\\and in large numbers on Sunday to vote in a historic referendum\\that will either remove left-wing President Hugo Chavez from\\office or give him a new mandate to govern for the next two\\years.',
 'S.Koreans Clash with Police on Iraq Troop Dispatch (Reuters) Reuters - South Korean police used water cannon in\\central Seoul Sunday to disperse at least 7,000 protesters\\urging the government to reverse a controversial decision to\\send more troops to Iraq.',
 "Palestinians in Israeli Jails Start Hunger Strike (Reuters) Reuters - Thousands of Palestinian\\prisoners in Israeli jails began a hunger strike for better\\conditions Sunday, but Israel's security minister said he\\didn't care if they starved to death."]

In [91]:
sports_ds = news_ds["train"].filter(lambda data: data["label"] == 1)
sports_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [92]:
business_ds = news_ds["train"].filter(lambda data: data["label"] == 2)
business_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [93]:
sci_tech_ds = news_ds["train"].filter(lambda data: data["label"] == 3)
sci_tech_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [94]:
news_ds["train"].features

{'text': Value(dtype='string', id=None),
 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}

In [95]:
label_map = news_ds["train"].features["label"].names

In [97]:
news_ds["train"].info

DatasetInfo(description='', citation='', homepage='', license='', features={'text': Value(dtype='string', id=None), 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}, post_processed=None, supervised_keys=None, builder_name='parquet', dataset_name='ag_news', config_name='default', version=0.0.0, splits={'train': SplitInfo(name='train', num_bytes=29832303, num_examples=120000, shard_lengths=None, dataset_name='ag_news'), 'test': SplitInfo(name='test', num_bytes=1880424, num_examples=7600, shard_lengths=None, dataset_name='ag_news')}, download_checksums={'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/train-00000-of-00001.parquet': {'num_bytes': 18585438, 'checksum': None}, 'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/test-00000-of-00001.parquet': {'num_bytes': 1234829, 'checksum': None}}, download_size=19820267, post_processing_size=None, dataset_size=31712727, size_in_bytes=51532994)

In [98]:
from pprint import pprint

pprint(news_ds["train"].info)

DatasetInfo(description='',
            citation='',
            homepage='',
            license='',
            features={'label': ClassLabel(names=['World',
                                                 'Sports',
                                                 'Business',
                                                 'Sci/Tech'],
                                          id=None),
                      'text': Value(dtype='string', id=None)},
            post_processed=None,
            supervised_keys=None,
            builder_name='parquet',
            dataset_name='ag_news',
            config_name='default',
            version=0.0.0,
            splits={'test': SplitInfo(name='test',
                                      num_bytes=1880424,
                                      num_examples=7600,
                                      shard_lengths=None,
                                      dataset_name='ag_news'),
                    'train': SplitInfo(name='train',
   

In [99]:
tokenizer("World", return_tensors="pt")
tokenizer("World", return_tensors="pt")["input_ids"].shape[-1]

2

In [100]:
label_sentences = [
    "Batman: This is mainly about the world, the one we’re sworn to protect.",
    "Joker: Why so serious? Clearly related to Sports news!",
    "Penguin: It’s all about Business, and trust me, I know how to play that game.",
    "Bane: This article mainly covers Sci/Tech.",
]

task_description = """
Task Description:
The scene unfolds with four iconic characters from the Batman universe seated around a table, each reading the same article. Batman, ever composed and logical, is the first to speak, delivering a clear and concise assessment of the piece. The Joker follows, making a mocking and unpredictable remark, turning the topic on its head. Two-Face, as always, relies on the flip of a coin to determine his take. Finally, Bane, with his focus on power and control, offers his perspective, emphasizing the article’s deeper implications.

Your task is simple: “Who explained the article the best?”.

Now, it’s time to choose—who described it best?

ARTICLE:"""

In [101]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.mps.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

In [102]:
news_ds["train"][1]["text"]

'Carlyle Looks Toward Commercial Aerospace (Reuters) Reuters - Private investment firm Carlyle Group,\\which has a reputation for making well-timed and occasionally\\controversial plays in the defense industry, has quietly placed\\its bets on another part of the market.'

In [103]:
news_ds["train"][1]["label"]

2

In [104]:
text_label = news_ds["train"][179]

probs = zero_shot_classification(
    text_label["text"],
    task_description,
    label_sentences,
)
print(probs)
import numpy as np

print(text_label["label"])
np.argmax(np.array(probs)).item()

[99.28125, 70.71484375, 158.25, 48.431640625]
3


2

In [86]:
def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_news_ds = news_ds.map(preprocess_function, batched=True)
tokenized_news_ds

Map:   0%|          | 0/120000 [00:00<?, ? examples/s]

Map:   0%|          | 0/7600 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 7600
    })
})

In [106]:
import numpy as np
from tqdm import tqdm
from colorama import Fore, Style, Back
import torch.nn.functional as F
import humanize
import math
import random


def calculate_nll(probs_softmax, label):
    return -torch.log(probs_softmax[label])


def get_confidence(probs_softmax):
    top_two_probs, _ = torch.topk(probs_softmax, 2)
    # 가장 높은 확률을 두 번째로 높은 확률로 나누어 비율을 계산
    return math.log10(top_two_probs[0] / top_two_probs[1])


total = 50
n_corrects = 0

# 0부터 7600 사이에서 겹치지 않게 50개의 숫자 뽑기
random_numbers = random.sample(range(7601), total)

for i in tqdm(range(total)):

    print(f"{Fore.YELLOW}----------------------------------{Fore.RESET}")
    text = tokenized_news_ds["test"][random_numbers[i]]["text"]
    label = tokenized_news_ds["test"][random_numbers[i]]["label"]
    probs = zero_shot_classification(
        text,
        task_description,
        label_sentences,
    )

    probs_tensor = torch.tensor(probs)
    probs_softmax = F.softmax(probs_tensor, dim=0)

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1
        tqdm.write(f"Index {i}: {Fore.GREEN}Success 🎉 {Style.RESET_ALL} ")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
    else:
        tqdm.write(f"Index {i}: {Fore.RED}Fail 🧐 {Style.RESET_ALL}")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
        print(f"Failed Query: {Back.YELLOW}{text}{Back.RESET}")
        label, pred = label_map[label], label_map[pred]
        print(
            f"Label : {Fore.CYAN} {label} {Fore.RESET} Pred : {Fore.MAGENTA} {pred} {Fore.RESET}"
        )
    current_accuracy = n_corrects / (i + 1)
    print(f"Current Accuracy: {Fore.BLUE}{current_accuracy:,.2%}{Fore.RESET}")

print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")
print(
    f"Total Correct Predictions: {Style.BRIGHT}{n_corrects} ({n_corrects/total:.2%}){Style.RESET_ALL}"
)
print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")

  0%|          | 0/50 [00:00<?, ?it/s]

----------------------------------


  2%|▏         | 1/50 [00:04<03:32,  4.34s/it]

Index 0: Fail 🧐 
dataset index: 6254
Confidence: 21.58
NLL: 49.69
Failed Query: Powell Meets with Middle East Leaders CBN.com - (CBN News) -ERUSALEM  -h the death of Yasser Arafat behind them, the Palestinians and Israelis are now facing an historic opportunity to move forward in peace.
Label :  World  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


  4%|▍         | 2/50 [00:06<02:38,  3.30s/it]

Index 1: Fail 🧐 
dataset index: 7412
Confidence: 28.53
NLL: inf
Failed Query: Last minute name change for Nvidia #39;s new GeForce 6200 with &lt;b&gt;...&lt;/b&gt; PCI Express allows Nvidia to tap into system memory to save expensive on-board graphics memory and achieve high performance at the same time.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


  6%|▌         | 3/50 [00:15<04:18,  5.50s/it]

Index 2: Fail 🧐 
dataset index: 2764
Confidence: 26.69
NLL: inf
Failed Query: .Mac bumps up storage capacity, improves mail (MacCentral) MacCentral - Apple has improved the services offered to subscribers of .Mac. Previously, the amount of storage for a basic .Mac account was 100MB, with a maximum of 15MB for e-mail. The service's base online storage has been increased to 250MB, e-mail service has been enhanced, and the cost of upgrading has been reduced. .Mac's basic subscription price remains the same -- US #36;99.95 per year.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


  8%|▊         | 4/50 [00:17<03:17,  4.30s/it]

Index 3: Fail 🧐 
dataset index: 2525
Confidence: 26.66
NLL: 61.38
Failed Query: Pinochet questioned by investigative judge An investigative judge has questioned former Chilean dictator Augusto Pinochet for half an hour to decide whether to indict him in one of hundreds of human rights cases stemming from his 1973-1990 rule.
Label :  World  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 10%|█         | 5/50 [00:20<02:45,  3.68s/it]

Index 4: Fail 🧐 
dataset index: 3733
Confidence: 30.02
NLL: 90.85
Failed Query: Lippi Wants Azzurri Improvement Italy boss Marcello Lippi is counting in his charges to make the country forget their weekend loss to Slovenia when they face Belarus in UEFA World Cup qualifying action on Wednesday.
Label :  Sports  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 12%|█▏        | 6/50 [00:22<02:22,  3.24s/it]

Index 5: Success 🎉  
dataset index: 7413
Confidence: 23.69
NLL: -0.00
Current Accuracy: 16.67%
----------------------------------


 14%|█▍        | 7/50 [00:25<02:11,  3.06s/it]

Index 6: Fail 🧐 
dataset index: 3965
Confidence: 26.01
NLL: 90.25
Failed Query: Like father, like son in Chrysler Classic If Brent Geiberger was pleased to win the Chrysler Classic of Greensboro, his father Al was positively ecstatic.  quot;I was going absolutely crazy watching it all unfold.
Label :  Sports  Pred :  Business 
Current Accuracy: 14.29%
----------------------------------


 16%|█▌        | 8/50 [00:35<03:49,  5.47s/it]

Index 7: Fail 🧐 
dataset index: 3546
Confidence: 24.51
NLL: 56.45
Failed Query: Study: Few Americans Buy Drugs Online NEW YORK - Only 4 percent of Americans have ever used the Internet to buy prescription drugs - and even fewer do so through foreign pharmacies - despite Web sites maintained by a handful of states to help citizens import medicines more cheaply from Canada, a new study finds.    A majority - 62 percent - believe drugs bought online are less safe than those purchased from a local pharmacy, accepting the federal government's stated concerns in opposing drug imports, the Pew Internet and American Life Project said in a report Sunday...
Label :  World  Pred :  Business 
Current Accuracy: 12.50%
----------------------------------


 18%|█▊        | 9/50 [00:38<03:04,  4.51s/it]

Index 8: Fail 🧐 
dataset index: 4495
Confidence: 28.38
NLL: 65.36
Failed Query: Zarqawi group claims attack on Australian soldiers in Baghdad (AFP) AFP - The group loyal to Al-Qaeda-linked militant Abu Mussab al-Zarqawi claimed to have bombed an Australian convoy in Baghdad, in a statement posted on an Islamist website.
Label :  World  Pred :  Business 
Current Accuracy: 11.11%
----------------------------------


 20%|██        | 10/50 [00:40<02:34,  3.86s/it]

Index 9: Fail 🧐 
dataset index: 6236
Confidence: 29.41
NLL: 84.79
Failed Query: Nine men Bhoys fall at Ibrox CELTIC lost ground in the title race today after Rangers won a heated Old Firm encounter that will be better remembered for its sending offs and scuffles.
Label :  Sports  Pred :  Business 
Current Accuracy: 10.00%
----------------------------------


 22%|██▏       | 11/50 [00:42<02:13,  3.41s/it]

Index 10: Fail 🧐 
dataset index: 5503
Confidence: 26.72
NLL: 88.15
Failed Query: Guard registers a special point Gary Payton didn't know he had joined the 20,000-point club last night until the public address announcer at the FleetCenter trumpeted the feat.
Label :  Sports  Pred :  Business 
Current Accuracy: 9.09%
----------------------------------


 24%|██▍       | 12/50 [00:45<01:57,  3.09s/it]

Index 11: Fail 🧐 
dataset index: 5421
Confidence: 25.33
NLL: inf
Failed Query: Stand Aside, Rudolph: The Mouse Will Lead A list of some of the best holiday gifts, based on taste, appearance or utility, available on the Web.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 8.33%
----------------------------------


 26%|██▌       | 13/50 [00:47<01:46,  2.89s/it]

Index 12: Success 🎉  
dataset index: 5252
Confidence: 24.88
NLL: -0.00
Current Accuracy: 15.38%
----------------------------------


 28%|██▊       | 14/50 [00:53<02:13,  3.72s/it]

Index 13: Fail 🧐 
dataset index: 1
Confidence: 25.70
NLL: inf
Failed Query: The Race is On: Second Private Team Sets Launch Date for Human Spaceflight (SPACE.com) SPACE.com - TORONTO, Canada -- A second\team of rocketeers competing for the  #36;10 million Ansari X Prize, a contest for\privately funded suborbital space flight, has officially announced the first\launch date for its manned rocket.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 14.29%
----------------------------------


 30%|███       | 15/50 [00:55<01:57,  3.37s/it]

Index 14: Fail 🧐 
dataset index: 1583
Confidence: 24.42
NLL: 83.79
Failed Query: Upsets Shake Up College Football Poll (AP) AP - The upsets have begun and the little guys are moving into The Associated Press poll. After ranked teams started the season 21-0, five fell to unranked opponents this weekend, shaking up media poll released Sunday.
Label :  Sports  Pred :  Business 
Current Accuracy: 13.33%
----------------------------------


 32%|███▏      | 16/50 [00:58<01:43,  3.05s/it]

Index 15: Fail 🧐 
dataset index: 7027
Confidence: 27.85
NLL: inf
Failed Query: Study: Return of Wolves Changes Ecosystem (AP) AP - Scientists studying the broader effects of wolf reintroduction said a growing body of evidence suggests that killing off predators such as wolves and grizzly bears in the last century started a cascade of effects that threw ecosystems out of balance.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 12.50%
----------------------------------


 34%|███▍      | 17/50 [01:03<02:00,  3.66s/it]

Index 16: Success 🎉  
dataset index: 1279
Confidence: 27.19
NLL: -0.00
Current Accuracy: 17.65%
----------------------------------


 36%|███▌      | 18/50 [01:05<01:39,  3.10s/it]

Index 17: Fail 🧐 
dataset index: 1065
Confidence: 28.60
NLL: 65.86
Failed Query: Scotch Whisky eyes Asian and Eastern European markets (AFP) AFP - A favourite tipple among connoisseurs the world over, whisky is treated with almost religious reverence on the Hebridean island of Islay, home to seven of Scotland's single malt distilleries.
Label :  World  Pred :  Business 
Current Accuracy: 16.67%
----------------------------------


 38%|███▊      | 19/50 [01:07<01:31,  2.95s/it]

Index 18: Success 🎉  
dataset index: 4613
Confidence: 25.53
NLL: -0.00
Current Accuracy: 21.05%
----------------------------------


 40%|████      | 20/50 [01:10<01:26,  2.88s/it]

Index 19: Fail 🧐 
dataset index: 658
Confidence: 26.62
NLL: 103.28
Failed Query: Dell adds new switch to lineup Dell has upgraded its PowerConnect line with the addition of the PowerConnect 5324, a 24-port managed gigabit layer 2 switch.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 20.00%
----------------------------------


 42%|████▏     | 21/50 [01:12<01:20,  2.76s/it]

Index 20: Fail 🧐 
dataset index: 6696
Confidence: 28.27
NLL: 100.51
Failed Query: Knicks: Crawford hits the one that counts The old axiom is that it doesn #39;t matter how many shots you miss; if you #39;re a shooter, you have to keep shooting. Jamal Crawford missed 21 shots against the Atlanta Hawks last 
Label :  Sports  Pred :  Business 
Current Accuracy: 19.05%
----------------------------------


 44%|████▍     | 22/50 [01:14<01:11,  2.54s/it]

Index 21: Success 🎉  
dataset index: 1924
Confidence: 26.23
NLL: -0.00
Current Accuracy: 22.73%
----------------------------------


 46%|████▌     | 23/50 [01:20<01:30,  3.37s/it]

Index 22: Fail 🧐 
dataset index: 798
Confidence: 26.98
NLL: 87.86
Failed Query: Britain Edges U.S. for 400M Relay Gold (AP) AP - Stymied by a sloppy handoff in the middle of the race, the United States lost to Great Britain by a hundredth of a second Saturday night in the 400-meter relay  #151; a race the American men usually dominate.
Label :  Sports  Pred :  Business 
Current Accuracy: 21.74%
----------------------------------


 48%|████▊     | 24/50 [01:22<01:20,  3.10s/it]

Index 23: Fail 🧐 
dataset index: 1876
Confidence: 27.92
NLL: 79.42
Failed Query: TV war puts tour of India in doubt Australian cricket chiefs fear a battle over television rights could cause next month #39;s Test series in India to be cancelled, and last night were seeking clarification from Indian board president 
Label :  Sports  Pred :  Business 
Current Accuracy: 20.83%
----------------------------------


 50%|█████     | 25/50 [01:25<01:14,  2.99s/it]

Index 24: Fail 🧐 
dataset index: 5799
Confidence: 26.07
NLL: 60.03
Failed Query: Timing Of Indian Move In Kashmir Vital: Pak Paper Islamabad, Nov. 16 (NNN): Pakistans leading newspaper, Dawn, finds the timing of the Indian announcement on reduction of troops in Kashmir as significance.
Label :  World  Pred :  Business 
Current Accuracy: 20.00%
----------------------------------


 52%|█████▏    | 26/50 [01:28<01:11,  2.97s/it]

Index 25: Fail 🧐 
dataset index: 5522
Confidence: 29.87
NLL: 88.59
Failed Query: Jet-Lagged Clarke Tied for Lead in Japan  GOTEMBA, Japan (Reuters) - Britain's Darren Clarke shook  off the effects of jet-lag and a hectic schedule to fire a  six-under-par 66 for a share of the first-round lead at the  Taiheiyo Masters Thursday.
Label :  Sports  Pred :  Business 
Current Accuracy: 19.23%
----------------------------------


 54%|█████▍    | 27/50 [01:30<01:02,  2.73s/it]

Index 26: Success 🎉  
dataset index: 2878
Confidence: 24.72
NLL: -0.00
Current Accuracy: 22.22%
----------------------------------


 56%|█████▌    | 28/50 [01:32<00:57,  2.62s/it]

Index 27: Fail 🧐 
dataset index: 2235
Confidence: 12.00
NLL: 61.97
Failed Query: Washington chooses a stadium site for Expos The DC Sports and Entertainment Commission outlined its plans Tuesday night in a meeting with city government officials. An official involved in the process, speaking on condition of anonymity, told The Associated 
Label :  Sports  Pred :  Business 
Current Accuracy: 21.43%
----------------------------------


 58%|█████▊    | 29/50 [01:35<00:55,  2.67s/it]

Index 28: Fail 🧐 
dataset index: 1313
Confidence: 27.47
NLL: 63.26
Failed Query: Iran ready to test Shahab-3 missile again: defense minister TEHRAN (IRNA) -- Defense Minister Ali Shamkhani stressed that Iran #39;s recent test of the Shahab-3 missile was successful, saying his ministry is ready to test it again  #39;in the presence of observers #39;.
Label :  World  Pred :  Business 
Current Accuracy: 20.69%
----------------------------------


 60%|██████    | 30/50 [01:38<00:53,  2.70s/it]

Index 29: Fail 🧐 
dataset index: 3186
Confidence: 25.23
NLL: 58.09
Failed Query: Tennis: Night final for Aus Open The centenary Australian Open will be the first Grand Slam event to stage its final at night.
Label :  World  Pred :  Business 
Current Accuracy: 20.00%
----------------------------------


 62%|██████▏   | 31/50 [01:40<00:48,  2.58s/it]

Index 30: Fail 🧐 
dataset index: 5183
Confidence: 29.21
NLL: 90.23
Failed Query: USC Releases Stewart From Scholarship (AP) AP - Southern California point guard Rodrick Stewart was granted a release from his basketball scholarship Friday.
Label :  Sports  Pred :  Business 
Current Accuracy: 19.35%
----------------------------------


 64%|██████▍   | 32/50 [01:43<00:47,  2.65s/it]

Index 31: Success 🎉  
dataset index: 7465
Confidence: 28.18
NLL: -0.00
Current Accuracy: 21.88%
----------------------------------


 66%|██████▌   | 33/50 [01:46<00:45,  2.66s/it]

Index 32: Fail 🧐 
dataset index: 2030
Confidence: 27.33
NLL: 62.92
Failed Query: Leaked secrets reveal chaos at heart of Blair #39;s Iraq plans TONY Blair was facing a new Iraq crisis last night after explosive evidence emerged from within his own government that he was warned the country would be plunged into chaos after the fall of Saddam Hussein.
Label :  World  Pred :  Business 
Current Accuracy: 21.21%
----------------------------------


 68%|██████▊   | 34/50 [01:48<00:40,  2.55s/it]

Index 33: Fail 🧐 
dataset index: 5608
Confidence: 29.77
NLL: 68.56
Failed Query: EU studies Iranian response on nuclear program wrangle VIENNA (AFP) - EU officials were evaluating Iran #39;s response to an offer for Tehran to avoid possible UN sanctions over its nuclear program in a wrangle that has led a UN watchdog to hold up a key report.
Label :  World  Pred :  Business 
Current Accuracy: 20.59%
----------------------------------


 70%|███████   | 35/50 [01:51<00:38,  2.54s/it]

Index 34: Fail 🧐 
dataset index: 862
Confidence: 27.14
NLL: 87.96
Failed Query: Angels' Glaus Activated From DL (AP) AP - Troy Glaus was activated from the 60-day disabled list Sunday by the Anaheim Angels and was back in the lineup against the Minnesota Twins.
Label :  Sports  Pred :  Business 
Current Accuracy: 20.00%
----------------------------------


 72%|███████▏  | 36/50 [01:53<00:36,  2.60s/it]

Index 35: Success 🎉  
dataset index: 4717
Confidence: 26.30
NLL: -0.00
Current Accuracy: 22.22%
----------------------------------


 74%|███████▍  | 37/50 [01:56<00:32,  2.52s/it]

Index 36: Fail 🧐 
dataset index: 3570
Confidence: 30.96
NLL: 71.29
Failed Query: Legendary all-rounder Miller dies Keith Miller, arguably Australia's greatest  all-rounder in Test cricket, has died in Melbourne aged 84.
Label :  World  Pred :  Business 
Current Accuracy: 21.62%
----------------------------------


 76%|███████▌  | 38/50 [01:59<00:31,  2.66s/it]

Index 37: Success 🎉  
dataset index: 2609
Confidence: 29.78
NLL: -0.00
Current Accuracy: 23.68%
----------------------------------


 78%|███████▊  | 39/50 [02:01<00:29,  2.70s/it]

Index 38: Success 🎉  
dataset index: 4056
Confidence: 27.73
NLL: -0.00
Current Accuracy: 25.64%
----------------------------------


 80%|████████  | 40/50 [02:04<00:26,  2.62s/it]

Index 39: Fail 🧐 
dataset index: 4248
Confidence: 22.83
NLL: 98.07
Failed Query: Briefly: Sun expands pay-as-you go supercomputing roundup Plus: SGI works on Linux performance software...Good Technology supported by HP, Samsung...RealNetworks loss widens on litigation.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 25.00%
----------------------------------


 82%|████████▏ | 41/50 [02:07<00:24,  2.75s/it]

Index 40: Fail 🧐 
dataset index: 5264
Confidence: 26.14
NLL: 101.89
Failed Query: VARs: Stone #39;s Exit Won #39;t Slow Novell #39;s Linux Drive Solution providers last week said they do not expect the sudden departure of Novell vice chairman Chris Stone, who engineered the company #39;s aggressive Linux push, to slow its Linux initiative.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 24.39%
----------------------------------


 84%|████████▍ | 42/50 [02:10<00:21,  2.73s/it]

Index 41: Fail 🧐 
dataset index: 2848
Confidence: 24.92
NLL: inf
Failed Query: Pharma Groups Work on EPC Issues Sept. 30, 2004Reacting to calls from pharmaceutical retailers, distributors and manufacturers, EPCglobal has added a new action group to specifically study the pharmaceutical industry 
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 23.81%
----------------------------------


 86%|████████▌ | 43/50 [02:12<00:18,  2.62s/it]

Index 42: Fail 🧐 
dataset index: 5885
Confidence: 27.52
NLL: 63.36
Failed Query: Doesn't matter to me if health care is privately delivered:Alta's Klein (Canadian Press) Canadian Press - EDMONTON (CP) - Alberta Premier Ralph Klein acknowledged Tuesday that he personally doesn't have a problem with private delivery of health-care services.
Label :  World  Pred :  Business 
Current Accuracy: 23.26%
----------------------------------


 88%|████████▊ | 44/50 [02:14<00:15,  2.57s/it]

Index 43: Fail 🧐 
dataset index: 6871
Confidence: 28.46
NLL: 84.26
Failed Query: Titans OT Munoz has surgery Knoxville, TN (Sports Network) - Tennessee Titans offensive tackle Michael Munoz underwent successful surgery Friday on his injured right shoulder.
Label :  Sports  Pred :  Business 
Current Accuracy: 22.73%
----------------------------------


 90%|█████████ | 45/50 [02:19<00:16,  3.26s/it]

Index 44: Fail 🧐 
dataset index: 15
Confidence: 21.72
NLL: inf
Failed Query: Teenage T. rex's monster growth Tyrannosaurus rex achieved its massive size due to an enormous growth spurt during its adolescent years.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 22.22%
----------------------------------


 92%|█████████▏| 46/50 [02:22<00:12,  3.12s/it]

Index 45: Fail 🧐 
dataset index: 5242
Confidence: 26.92
NLL: 81.25
Failed Query: DaLomba sprints to win With the Eastern Massachussetts cross-country championships just a week away, yesterday's MSTCA Invitational at Franklin Park offered area runners a last chance to tune up for the title race.
Label :  Sports  Pred :  Business 
Current Accuracy: 21.74%
----------------------------------


 94%|█████████▍| 47/50 [02:25<00:09,  3.00s/it]

Index 46: Success 🎉  
dataset index: 1093
Confidence: 27.59
NLL: -0.00
Current Accuracy: 23.40%
----------------------------------


 96%|█████████▌| 48/50 [02:28<00:05,  2.96s/it]

Index 47: Fail 🧐 
dataset index: 5058
Confidence: 23.17
NLL: 82.03
Failed Query: Roma 1-1 Bayer Leverkusen: FT Report ROME, November 3 (Champions League) - Vincenzo Montella #39;s injury-time equaliser forced Bayer Leverkusen to settle for a share of the points on Wednesday in Group B of the Champions League but the Eternal City club are virtually eliminated if not yet 
Label :  Sports  Pred :  Business 
Current Accuracy: 22.92%
----------------------------------


 98%|█████████▊| 49/50 [02:30<00:02,  2.84s/it]

Index 48: Fail 🧐 
dataset index: 4456
Confidence: 16.99
NLL: 39.11
Failed Query: Victory Looms for Karzai as Vote Probe Continues Hamid Karzai was assured of a majority in Afghanistans election to become its first democratically chosen president. With nearly 95 per cent of votes counted, the interim leader already has more than half 
Label :  World  Pred :  Business 
Current Accuracy: 22.45%
----------------------------------


100%|██████████| 50/50 [02:33<00:00,  3.06s/it]

Index 49: Fail 🧐 
dataset index: 2414
Confidence: 27.29
NLL: inf
Failed Query: Samples From Genesis Craft Sent to Calif. (AP) AP - The first solar-wind samples recovered from the crashed Genesis space capsule have been sent to researchers in California.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 22.00%
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Total Correct Predictions: 11 (22.00%)
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>


---


# 실습: Zero-shot Classification

이번 실습에서는 open LLM을 가지고 zero-shot classification을 해봅니다. 먼저 필요한 library들을 설치합시다.


그 다음 Gemma-2B를 사용하기 위해 다음과 같은 작업을 진행합니다:

1. huggingface.co 계정 만들고 로그인하기
2. https://www.kaggle.com/models/google/gemma/license/consent 에서 Gemma license 동의하기
3. 홈 화면으로 돌아와, `Profile > Settings > Access Tokens` 메뉴로 들어와 "Write" type의 token 생성하기
4. 생성한 토큰을 아래 "HF TOKEN"에 불여넣고 셀을 실행하기.


In [ ]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")

이번에는 Gemma-2B를 가지고 간단한 text 생성을 해봅시다.
"What is your name?" 이라는 text를 넣었을 때 어떤 text가 생성되는지 살펴봅시다.


In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
input_ids

In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
outputs = model.generate(**input_ids)
print(tokenizer.decode(outputs[0]))

In [ ]:
tokens = input_ids["input_ids"]
print(tokens)
input_ids

2B의 작은 LLM이라 질좋은 답변이 나오지 않는 것을 알 수 있습니다.
이번에는 입력으로 넣어준 token들의 logit을 계산해봅시다.


In [ ]:
tokens = input_ids["input_ids"]
print(tokens)

logits = model(**input_ids).logits
for i in range(tokens.shape[-1]):
    token = tokens[0, i].item()
    print(logits[0, i, token])

위와 같이 모델 출력의 `.logits`을 통해 token들의 logit을 알 수 있습니다.
Logit은 높을 수록 token이 나올 확률이 높다는 뜻입니다.

이번에는 logit 계산을 통해 zero-shot classification을 구현해보도록 하겠습니다.


In [8]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.cuda.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

아래는 실제로 zero-shot classification을 해본 결과입니다.


In [ ]:
probs = zero_shot_classification(
    "I am happy!", "Is the sentence positive or negative?: ", ["positive", "negative"]
)
print(probs)

보시다시피 우리는 Gemma를 별도로 학습하지 않았음에도 불구하고 주어진 문장이 긍정적이라는 것을 정확하게 예측하고 있습니다.

다음은 영화 리뷰 감정 분석 task에 적용해봅시다.
먼저 data를 불러옵니다.


In [ ]:
from datasets import load_dataset


imdb = load_dataset("imdb")


def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_imdb = imdb.map(preprocess_function, batched=True)

그리고 `test` data에서 50개의 영화 리뷰에 대해 예측하는 코드는 다음과 같습니다.


In [ ]:
import numpy as np
from tqdm import tqdm


n_corrects = 0
for i in tqdm(range(50)):
    text = tokenized_imdb["test"][i]["text"]
    label = tokenized_imdb["test"][i]["label"]
    probs = zero_shot_classification(
        text,
        "A movie review is given. Decide that the movie review is positive or negative: ",
        labels=["Answer: negative.", "Answer: positive."],
    )

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1

print(n_corrects)

In [ ]:
text = tokenized_imdb["test"][i]["text"]
text

In [ ]:
label = tokenized_imdb["test"][i]["label"]
label

보시다시피 정확도 88%로, 매우 높은 성능을 보이는 것을 알 수 있습니다.
